pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,287 @@
+	if !kubetypes.IsCriticalPod(pod) || !utilfeature.DefaultFeatureGate.Enabled(features.ExperimentalCriticalPodAnnotation) {
+		return false, failureReasons, nil
+	}
+	// InsufficientResourceError is not a reason to reject a critical pod.
+	// Instead of rejecting, we free up resources to admit it, if no other reasons for rejection exist.
+	nonResourceReasons := []algorithm.PredicateFailureReason{}
+	resourceReasons := []*admissionRequirement{}
+	for _, reason := range failureReasons {
+		if r, ok := reason.(*predicates.InsufficientResourceError); ok {
+			resourceReasons = append(resourceReasons, &admissionRequirement{
+				resourceName: r.ResourceName,
+				quantity:     r.GetInsufficientAmount(),
+			})
+		} else {
+			nonResourceReasons = append(nonResourceReasons, reason)
+		}
+	}
+	if len(nonResourceReasons) > 0 {
+		// Return only reasons that are not resource related, since critical pods cannot fail admission for resource reasons.
+		return false, nonResourceReasons, nil
+	}
+	err := c.evictPodsToFreeRequests(admissionRequirementList(resourceReasons))
+	// if no error is returned, preemption succeeded and the pod is safe to admit.
+	return err == nil, nil, err
+}
+
+// freeRequests takes a list of insufficient resources, and attempts to free them by evicting pods
+// based on requests.  For example, if the only insufficient resource is 200Mb of memory, this function could
+// evict a pod with request=250Mb.
+func (c *CriticalPodAdmissionHandler) evictPodsToFreeRequests(insufficientResources admissionRequirementList) error {
+	podsToPreempt, err := getPodsToPreempt(c.getPodsFunc(), insufficientResources)
+	if err != nil {
+		return fmt.Errorf("preemption: error finding a set of pods to preempt: %v", err)
+	}
+	glog.Infof("preemption: attempting to evict pods %v, in order to free up resources: %s", podsToPreempt, insufficientResources.toString())
+	for _, pod := range podsToPreempt {
+		status := v1.PodStatus{
+			Phase:   v1.PodFailed,
+			Message: message,
+			Reason:  reason,
+		}
+		// record that we are evicting the pod
+		c.recorder.Eventf(pod, v1.EventTypeWarning, reason, message)
+		// this is a blocking call and should only return when the pod and its containers are killed.
+		err := c.killPodFunc(pod, status, nil)
+		if err != nil {
+			return fmt.Errorf("preemption: pod %s failed to evict %v", format.Pod(pod), err)
+		}
+		glog.Infof("preemption: pod %s evicted successfully", format.Pod(pod))
+	}
+	return nil
+}
+
+// getPodsToPreempt returns a list of pods that could be preempted to free requests >= requirements
+func getPodsToPreempt(pods []*v1.Pod, requirements admissionRequirementList) ([]*v1.Pod, error) {
+	bestEffortPods, burstablePods, guaranteedPods := sortPodsByQOS(pods)
+
+	// find the guaranteed pods we would need to evict if we already evicted ALL burstable and besteffort pods.
+	guarateedToEvict, err := getPodsToPreemptByDistance(guaranteedPods, requirements.subtract(append(bestEffortPods, burstablePods...)...))
```

Comment:
We are walking over every single pod in this algorithm. What if we make it greedy and just choose a victim with has a distance approximately equal to 0. Given that we are disrupting pods anyways, I don't know if the additional scrutiny is warranted. From a user's perspective, they cannot predict which pod will get preempted no matter which option we choose.

1. Check if any `besteffort` pod would have a distance ~0.
2. `1` can be true only if the resource that's not available is `cpu` or `memory`, but it is possible.
3. If `besteffort` pods don't meet the requirements, then find `Burstable` pods to preempt.
4. If `3` isn't enough, identify guaranteed pods.

The distance delta can be `0.1` for example. Such an algorithm will be greedy, but not necessarily accurate or predictable. WDYT? 

Again, if we had metrics around this, it might help inform this conversation.
Given the time pressure, we can make the change as an optimization later too. Having metrics around preemption time would help though, because that will contribute to pod startup delays.